Implement Positioned, Vectored IO in the Page Cache

chrisvest opened the pull request:
This significantly speeds up the store flushing. Especially on fast storage devices.
The reason vectored IO speeds up our writes, is our odd-sized file pages.
Because the file pages are not sized as a multiple of the OS pages, the OS needs to read in the last block in a page, so that most of the data can be overwritten by the write, and then the whole block - included the last few unchanged bytes - can be written back to the device.
It is this read which slows us down, because we cannot make any progress until it completes.
This defeats many buffering and asynchronisity optimisations on many layers down to the device.
With the vectored IO methods, the last bytes of those blocks can now be found in the next page of the vector, which eliminates all but one read.
This allows us to stream data to the storage device in chunks of up to 32 MiB at a time, easily saturating all write buffers of the system, with the exception of the OS page cache.
The net effect is that we now stand a good chance of writing at the maximum speed of the underlying storage subsystem.

chrisvest committed:
Add the necessary APIs to support positioned vectored IO through the PageSwappers.

Note that at present, the support is faked with non-vectored positioned writes.
Also, the page cache does not yet make use of these new APIs.

The idea is that the operating system will be able to combine the writes in its page cache, and avoid the read-modify-write situations we otherwise see because of our odd file page sizes.

chrisvest committed:
Change the SingleFilePageSwapperFactory implementation name from striped to single, because we might want to stripe over files later

chrisvest committed:
Implement positioned, vectored IO in the SingleFilePageSwapper

And make the PagedFile.flushAndForce and PageCache.flushAndForce methods use it.
The background flush thread and the eviction thread are still flushing page by page with pwrite.

This significantly speeds up the store flushing. Especially on fast storage devices.
The reason vectored IO speeds up our writes, is our odd-sized file pages.
Because the file pages are not sized as a multiple of the OS pages, the OS needs to read in the last block in a page, so that most of the data can be overwritten by the write, and then the whole block - included the last few unchanged bytes - can be written back to the device.
It is this read which slows us down, because we cannot make any progress until it completes.
This defeats many buffering and asynchronisity optimisations on many layers down to the device.
With the vectored IO methods, the last bytes of those blocks can now be found in the next page of the vector, which eliminates all but one read.
This allows us to stream data to the storage device in chunks of up to 32 MiB at a time, easily saturating all write buffers of the system, with the exception of the OS page cache.
The net effect is that we now stand a good chance of writing at the maximum speed of the underlying storage subsystem.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/PageSwapper.java:
```diff
@@ -43,9 +43,28 @@
      * channel and the operation must be retried.
      */
+    // TODO all these read and write methods should return long instead of int
     int read( long filePageId, Page page ) throws IOException;
```
That TODO would be a good thing to complete before merging.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPagedFile.java:
```diff
@@ -184,25 +189,40 @@ public void flushAndForce() throws IOException
             {
+                // TODO only flush pages that are actually dirty, except we can tolerate flushing a few clean pages if
+                // TODO it means we can use larger vectors
+                int pagesGrabbed = 0;
```
This also needs to be investigated more.

chrisvest commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPagedFile.java:
```diff
@@ -214,6 +234,43 @@ void flushAndForceInternal( FlushEventOpportunity flushOpportunity ) throws IOEx
+                pages[j].unlockRead( stamps[j] );
+            }
+        }
+    }
```
`MuninnPage` has a `flush()` method, but can't get a corresponding vectored version. This creates a weird design where some IO is done through the `MuninnPage` object, while other IO code is entirely contained in the `PageSwapper`. This should be sorted out and made consistent.

chrisvest committed:
Make the PageSwapper read and write methods return the byte counts as longs instead of ints

This is necessary now that the vectored methods can cover more than one page, though it won't make any difference in practice.

The non-vectored methods also return long byte counts for consistency.

chrisvest committed:
Gently add comments to the new vectored flushing code in MuninnPagedFile

chrisvest committed:
Fix a bug where vectored flushing can race with eviction

Previously we iterated the translation table and added all pages to our IO vector, that we both loaded and dirty.
However, this is not a sufficiently strict criteria, beucase in between reading the page out of the translation table and locking it, it can be evicted and faulted into another page.
If this happened, we could end up adding a dirty page *from another file* to our IO vector, thus mixing the contents of the file.

To fix this, I have changed the check for whether the page is loaded or not, to a check for whether the page is bound to the expected file and file page id.
The flushing code then computes what file page id we should expect to see, from the location of the page in the translation table.
This also means that we won't be fooled if the page is faulted into the same file but at a different location.

I have added no tests for this, because the data race is too narrow to hit with a targeted test. This bug was found through code inspection.

chrisvest committed:
Make the parallel batch importer use less page cache memory

Also make the importer do sequential, vectored IO by making it flush the stores after each batch.

chrisvest committed:
Add page cache test for rare race bug

Add a test for the case where vectored flushing can race with eviction and page faulting, such that the flusher reads a page out of its translation table,
but before it can grab a read lock on that page, the page gets evicted and then faulted into a different swapper+filePageId combination and marked as dirty.
Previously, we would only check if a page was loaded before adding it to the vector of pages to be written, but now we check that the page is *bound* to the
expected swapper and file page id, and we do this check after we've grabbed the read lock on the page. This prevents the aforementioned race from happening.

This fix was already committed in #f64422b5487cfcd36f7acce0e1a970043d54e4ff - in this commit, we've modified the
concurrentFlushingMustNotPutInterleavedDataIntoFile test, such that it is capable of catching this bug.

chrisvest committed:
Fix a pread bug in the EphemeralFSA

The bug was that a positioned read of an empty file did not correctly recognise that we were reading beyond the end of the file.
So, instead of returning -1 as the EOF signal, it would return the negative of the requested file position, which is complete non-sense.
The fix is that we now check for any figure of available bytes that is less than or equal to zero, and return the EOF signal for that.
The old code required that the number of available bytes ended up as exactly zero, which of course is quite rare, since most requested
file positiones are not exactly zero.

MishaDemianenko commented on community/import-tool/src/main/java/org/neo4j/tooling/ImportTool.java:
```diff
@@ -415,15 +415,15 @@ private static void validateInputFiles( Collection<Option<File[]>> nodesFiles,
+            public int writeBufferSize()
             {
-                return processors != null ? processors.intValue() : super.maxNumberOfProcessors();
+                return defaultSettingsSuitableForTests? 1024 * 1024 * 8 : super.writeBufferSize();
```
Looks like magic number to me, reasonable name can add a bit of clarity here

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/fs/StoreFileChannel.java:
```diff
@@ -170,4 +170,10 @@ public void flush() throws IOException
         force( false );
     }
+
+    public static FileChannel unwrap( StoreChannel channel )
```
Why not just simple instance method?

chrisvest replied:
Because it's not part of the "normal" API. It is a method that circumvents the encapsulation of the StoreFileChannel.

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/SingleFilePageSwapper.java:
```diff
@@ -308,7 +433,85 @@ public int write( long filePageId, Page page ) throws IOException
+    {
+        try
+        {
+            synchronized ( positionLock( channel ) )
```
What about instead of unwrap, retrieving lock each time we modify `StoreFileChannel` abstraction to provide it for us? then we do not need those unwraps and can cache lock for a storeFileChannel lifetime?

chrisvest replied:
No, the concept of a position lock is specific to `sun.nio.ch.FileChannelImpl`. The StoreFileChannel has no idea what concrete type of FileChannel it is wrapping. Grabbing it is a hack, that I don't want to spread. This is the only place in our code that should have a need for doing this.

chrisvest committed:
Move the StoreFileChannel.unwrap method to a dedicated class

This way, the mechanism for bypassing the encapsulation of the StoreFileChannel is further removed from the StoreFileChannel itself.
In other words, the lock and the key are put further apart.

chrisvest committed:
Pull constant expression into constant field

The write buffer size is used for configuring the page cache memory allocation, and we usually give the page cache 8 MiB to play with when we test.
